sql/parser/eval.go

```diff
@@ -650,25 +650,28 @@ func evalUnaryExpr(expr *UnaryExpr) (Datum, error) {
 
 func evalFuncExpr(expr *FuncExpr) (Datum, error) {
 	name := strings.ToLower(expr.Name.String())
-	b, ok := builtins[name]
-	if !ok {
-		return DNull, fmt.Errorf("%s: unknown function", expr.Name)
-	}
-	if b.nArgs != -1 && b.nArgs != len(expr.Exprs) {
-		return DNull, fmt.Errorf("%s: incorrect number of arguments: %d vs %d",
-			expr.Name, b.nArgs, len(expr.Exprs))
-	}
-
 	args := make(DTuple, 0, len(expr.Exprs))
-	for _, e := range expr.Exprs {
+	var types [5]string
+	for i, e := range expr.Exprs {
 		arg, err := EvalExpr(e)
 		if err != nil {
 			return DNull, err
 		}
 		args = append(args, arg)
+		types[i] = arg.Type()
+	}
+
+	sig := builtin{
+		name:  name,
+		types: types,
+	}
+
+	fn, ok := builtins[sig]
```

Comment:
We should file an issue that we shouldn't be performing function lookup every time `EvalExpr` is called. Like `ValArg` and `QualifiedName`, we should perform a pass over the expression and resolve function names to their implementations. Might also be possible to verify types at that time.